iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -320,14 +322,102 @@ static Value buildFlowWorkgroupInfoOp(OpBuilder &b, unsigned dim) {
+  // For each operation compute the list of operations in `ops` that use its
+  // results. Also compute the operations that form the leafs of the DAG of
+  // operations in `ops`.
+  for (auto op : ops) {
+    for (auto operand : op->getOperands()) {
+      auto definingOp = operand.getDefiningOp();
+      if (!definingOp) continue;
+      insertAfterMap[definingOp].push_back(op);
+      if (leafOps.count(definingOp)) leafOps.remove(op);
+    }
+  }
+
+  // The leaves are at the head of the ordered list.
+  SmallVector<Operation *, 16> orderedOps = llvm::to_vector<16>(leafOps);
+  orderedOps.reserve(ops.size());
+  llvm::SmallPtrSet<Operation *, 16> processed;
+  processed.insert(leafOps.begin(), leafOps.end());
+
+  // `readyOps` contains the list of operations that have been just added to the
+  // `orderedOps` list. With these marked ready, they might make further
+  // operations in `ops` ready as well.
+  // The complexity of the algorithm is driven by these
+  // - Each operations is added to `readyOps` list at most once, and is removed
+  //   after being processed
+  // - For every operation in `readyOps` every use of its results (within `ops`)
+  //   is looked at once.
+  // - For every use, the operands of the user are processed.
+  // Assuming operands is O(1), i.e. constant order, the complexity is O(sum of
+  // number of uses of each operation). Given that the size of `ops` is at max
+  // O(10), and not O(100), this is assumed to be reasonable.
+  SmallVector<Operation *, 16> readyOps = llvm::to_vector<16>(leafOps);
+  while (!readyOps.empty()) {
+    SmallVector<Operation *> nextReadyOps;
+    DEBUG_WITH_TYPE(DEBUG_TYPE, {
+      llvm::dbgs() << "ReadyOps :\n";
+      for (auto readyOp : readyOps) {
+        llvm::dbgs() << "\t";
+        readyOp->print(llvm::dbgs());
+        llvm::dbgs() << "\n";
+      }
+    });
+    for (auto readyOp : readyOps) {
+      // Check all uses of `readyOp` within `ops`. If all of the operations that
+      // define the operands of the user have been added to `orderedOps`, then
+      // the user is ready to be scheduled.
+      for (auto insertAfterOp : insertAfterMap[readyOp]) {
+        if (processed.count(insertAfterOp)) continue;
+        if (llvm::all_of(insertAfterOp->getOperands(), [&](Value operand) {
+              Operation *operandDefiningOp = operand.getDefiningOp();
+              return !operandDefiningOp || processed.count(operandDefiningOp);
+            })) {
+          nextReadyOps.push_back(insertAfterOp);
+          orderedOps.push_back(insertAfterOp);
+          processed.insert(insertAfterOp);
+        }
+      }
+    }
+    std::swap(nextReadyOps, readyOps);
+  }
+
+  DEBUG_WITH_TYPE(DEBUG_TYPE, {
+    llvm::dbgs() << "Ops to be inlined (sorted) : \n";
+    for (auto op : orderedOps) {
+      llvm::dbgs() << "\t";
+      op->print(llvm::dbgs());
+      llvm::dbgs() << "\n";
+    }
+  });
+  assert(orderedOps.size() == ops.size());
+  return std::move(orderedOps);
```

Comment:
I think we don't need `std::move` in this case. I learned that this is not needed, and it will prevent the compiler to use return-value-optimization.

https://stackoverflow.com/questions/44065808/returning-stdvector-with-stdmove
https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization

Some snippets:

> Whenever a local variable in a return statement is eligible for copy elision, it binds to an rvalue re­fe­rence, and thus `return t`; is identical to `return std::move(t);`.

> using an explicit std::move as this will prevent the compiler to use RVO.

Reply:
Issue was i was trying to return a `SmallVector<Operation *, 16>` to a `SmallVector<Operation *>`. Warning on compiler threw me off. Removed this now.